src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java

```diff
@@ -20,45 +20,171 @@
+            results.add(_expression.next());
+            FlworTuple newTuple = new FlworTuple(_variableName, results);
+            _nextLocalTupleResult = newTuple;
+            this._hasNext = true;
+        } else {
+            _expression.close();
+            this._hasNext = false;
+        }
+    }
+
+    @Override
+    public void open(DynamicContext context) {
+        super.open(context);
+
+        // isRDD checks omitted, as open is used for non-RDD(local) operations
+
+        if (this._child != null) { //if it's not a start clause
+            _child.open(_currentDynamicContext);
+            _tupleContext = new DynamicContext(_currentDynamicContext);     // assign current context as parent
+
+            setNextLocalTupleResult();
+
+        } else {    //if it's a start clause
+            _expression.open(this._currentDynamicContext);
+
+            initialFor_setNextLocalTupleResult();
+        }
+    }
+
+    @Override
+    public void close() {
+        if (_child != null) {
+            _child.close();
+        }
     }
 
 
     @Override
     public JavaRDD<FlworTuple> getRDD() {
-        if (this._rdd == null) {
-            String variableReference = ((VariableReferenceIterator)this._children.get(0)).getVariableName();
-            RuntimeIterator assignmentExpression = this._children.get(1);
+        if (this._rdd == null) {        //if it's a start clause
             JavaRDD<Item> initialRdd = null;
-            //if it's a start clause
+
             if (this._child == null) {
-                initialRdd = this.getNewRDDFromExpression(assignmentExpression);
-                this._rdd = initialRdd.map(new InitialForClauseClosure(variableReference));
-            } else {
-            //if it's not a start clause
-                this._rdd = this._child.getRDD();
-                this._rdd = this._rdd.flatMap(new ForClauseClosure(assignmentExpression, variableReference));
+                initialRdd = this.getNewRDDFromExpression(_expression);
+                this._rdd = initialRdd.map(new InitialForClauseClosure(_variableName));
+            } else {        //if it's not a start clause
+
+                if (_child.isRDD()) {
+                    this._rdd = this._child.getRDD();
+                    this._rdd = this._rdd.flatMap(new ForClauseClosure(_expression, _variableName));
+
+                } else {    // if child is locally evaluated
+                    // _expression is definitely an RDD if execution flows here
+
+                    // get child tuples
+                    List<FlworTuple> localTuples = new ArrayList<>();
+                    _child.open(_currentDynamicContext);
+                    while (_child.hasNext()) {
+                        localTuples.add(_child.next());
+                    }
+                    _child.close();
+
+                    // turn them into an rdd
+                    _rdd = SparkContextManager.getInstance().getContext().parallelize(localTuples);
```

Comment:
Here you are basically materializing the (potentially big) expression and turning a (usually small because not an RDD) set of tuples into an RDD.

I think it can/should be done the other way round, i.e., local iteration on the input tuples, and for each input tuple, ask for the corresponding output tuple RDD (obtained from a map on the expression RDD. This is a new closure class that maps each item I of the expression to a tuple that is a copy of the input tuple, plus a binding from the variable to I, with the tuple constructor you already created). That gives you one RDD per tuple, and then you can call a union on all these RDD to obtain the final output RDD.

The use case is queries such as
for $i in (1000, 10000, 1000000)
for $j in parallelize(1 to $i)
return...

or

for $f in ("file1.json", "file2.json")
for $i in json-file($f)
return ...

Sorry for having not been precise enough yesterday! I think the case we discussed was a special case in which the expression doesn't depend on the input tuple.

Reply:
I have just pushed an implementation try at this. The current forClauseClosure creates a single tuple for each result obtained from the expression. The created tuple is a duplicate of the input tuple containing the new variable binding. I believe this is exactly what you meant in your description so I thought the existing implementation can be used here.

In its current state operation is as follows:
1-) read a single child tuple
2-) Turn it into an RDD with parallelize (not sure if this is the best way)
3-) Apply the forclosure as explained above
4-) union all RDDs(different one for each child tuple) at the end

Is there an implementation detail I'm missing/ have misunderstood?

Reply:
Almost :-)

The map shouldn't map a tuple to tuples. It should map an *item* (from the expression RDD) to tuples. This is a new closure class.

Here is a simple example:

for $i in (2, 3)
for $j in parallelize(1 to $i)
return $i + $j

the two input tuples to the second for clause (locally available) are
(i: 2)
(i:3)

what needs to be done is to iterate through these two tuples locally (without creating any RDD from any of them):

- for the tuple (i:2), take the RDD returned by the expression (internally contains 1, 2)

map that rdd of items to a tuple RDD like so:

1 -> mapped to (i:2, j:1) 
2 -> mapped to (i:2, j:2) 

This is a mapping from an RDD of items to an RDD of tuples. It is a closure that has a private variable containing the input tuple T (here T = (i:2)) passed through the constructor.
It takes x and returns (T, j: x)

The output RDD for that input tuple is thus made of:
(i:2, j:1) 
(i:2, j:2) 

- for the tuple (i:3), take the RDD returned by the expression (internally contains 1, 2, 3)

1 -> mapped to (i:3, j:1) 
2 -> mapped to (i:3, j:2) 
2 -> mapped to (i:3, j:2) 

This is the same closure, but with T = (i:3).

The output RDD for that input tuple is thus made of:
(i:3, j:1) 
(i:3, j:2) 
(i:3, j:3)

- finally, you union all output rdds:
 (i:2, j:1) 
(i:2, j:2) 
(i:3, j:1) 
(i:3, j:2) 
(i:3, j:3)

and output this RDD as the output of the second clause.

Does it help?

Reply:
Thank you so much for the detailed explanation. I realized that I understood it backwards in a sense in the first place. I have pushed another implementation just now. Unfortunately in its current state I keep facing a "task not serializable" exception and I haven't been able to figure out the cause yet. I'll keep looking for the source of the problem. In the mean time, if you catch something that I'm missing please let me know

Reply:
Very good. The task not serializable is issue is, in all likelihood, a class that doesn't bear the "Serializable" interface, in the transitive closure of all the data stored in the closure class fields.

Reply:
Through debugging I extracted the following exception cause: It seems FlworTuple is causing the issue. However, the class is KryoSerializable.

java.io.NotSerializableException: sparksoniq.jsoniq.tuple.FlworTuple
Serialization stack:
	- object not serializable (class: sparksoniq.jsoniq.tuple.FlworTuple, value: sparksoniq.jsoniq.tuple.FlworTuple@760f1081)
	- field (class: sparksoniq.spark.closures.ForClauseLocalToRDDClosure, name: _inputTuple, type: class sparksoniq.jsoniq.tuple.FlworTuple)
	- object (class sparksoniq.spark.closures.ForClauseLocalToRDDClosure, sparksoniq.spark.closures.ForClauseLocalToRDDClosure@7a358613)
	- field (class: org.apache.spark.api.java.JavaPairRDD$$anonfun$toScalaFunction$1, name: fun$1, type: interface org.apache.spark.api.java.function.Function)
	- object (class org.apache.spark.api.java.JavaPairRDD$$anonfun$toScalaFunction$1, <function1>)

Reply:
Exactly! :-)

Reply:
Good news! I have managed to work through the serialization issue and the code is functional now. I have also added the case you've mentioned above as a test case.